Run performance benchmark on pull requests

rishabh6788 opened the pull request:
### Description
Coming from #14553. 
This PR adds 2 new github actions workflows:
1. `benchmark-pull-request`: This workflow is initiated when a user adds a comment of the format `{"run-benchmark-test":"<config-id>"` on pull request. The workflow then builds the pull request code, uploads artifact to an S3 bucket, uses that artifact to spin up a cluster (single or multi-node), runs benchmark using [opensearch-benchmark](https://github.com/opensearch-project/opensearch-benchmark) and finally updates the result on the pull request. Please refer the `PERFORMANCE_BENCHMARKS.md` readme for details on how it works. 
2. `add-performance-comment`: This simple workflow runs whenever a user adds `Performance` label to the PR and adds a comment to the PR informing user about the performance benchmark workflow and how to run it, if interested. 

The related jobs and jenkins changes have already been completed or in-review. Related PRs:
* https://github.com/opensearch-project/opensearch-build/pull/4846
* https://github.com/opensearch-project/opensearch-build/pull/4831

Here is the sample PR with results https://github.com/rishabh6788/OpenSearch/pull/5. 
Sample workflow run https://github.com/rishabh6788/OpenSearch/actions/runs/9945672578/job/27474462658. 

### Related Issues
#14553 

### Check List
- ~[ ] Functionality includes testing.~
- ~[ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.~
- ~[ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

rishabh6788 commented:
Adding @reta @dblock @getsaurabh02 @peternied @msfroh @sohami @backslasht @andrross for comments and feedback.

reta commented on .github/workflows/benchmark-pull-request.yml:
```diff
@@ -0,0 +1,145 @@
+              issue_number: context.issue.number,
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: 'Invalid comment format or config id.'
```
May be just hint what the correct format should be?

```suggestion
              body: 'Invalid comment format or config id. Use `run-benchmark-test <id>`'
```

reta replied:
(or link to docs)

reta commented on .github/workflows/benchmark-pull-request.yml:
```diff
@@ -0,0 +1,145 @@
+name: Run performance benchmark on pull request
+on:
+  issue_comment:
```
This part confuses me: the benchmark run is on pull request (make sense) but comment is on the issue (confusion)? I would expect it to be on the pull request, where the code lives. Plus, the issue could have several pull requests "associated" with it, fe multiple  backports or large features split in chunks.

reta commented on PERFORMANCE_BENCHMARKS.md:
```diff
@@ -0,0 +1,112 @@
+
+## Overview
+
+`benchmark-pull-request` GitHub Actions workflow is designed to automatically run performance benchmarks on a pull request when a specific comment is made on the related issue. This ensures that performance benchmarks are consistently and accurately applied to code changes, helping maintain the performance standards of the repository.
```
I would expect the comment to be on the pull request, where the code lives. Plus, the issue could have several pull requests "associated" with it, fe multiple backports or large features split in chunks.

rishabh6788 replied:
Sorry for the confusion, the comment has to be added on the pull request and not on any issue. Will correct the wording.

reta replied:
Could we please updated docs accordingly?

rishabh6788 replied:
Done!

peterzhuamazon commented:
Might want to limit the number of concurrently runs like we did with gradle check.
To avoid too many runs initiated through PR comments:
https://github.com/opensearch-project/opensearch-build/blob/main/jenkins/gradle/gradle-check.jenkinsfile#L20-L27

Thanks.

rishabh6788 commented:
> Might want to limit the number of concurrently runs like we did with gradle check. To avoid too many runs initiated through PR comments: https://github.com/opensearch-project/opensearch-build/blob/main/jenkins/gradle/gradle-check.jenkinsfile#L20-L27
> 
> Thanks.

This is a good point, will update the job in another PR.

reta commented on .github/workflows/benchmark-pull-request.yml:
```diff
@@ -0,0 +1,145 @@
+          role-to-assume: ${{ secrets.UPLOAD_ARCHIVE_ARTIFACT_ROLE }}
+          role-session-name: publish-to-s3
+          aws-region: us-west-2
+      - name: Push to S3
```
I believe we need a S3 cleanup step after `Trigger jenkins workflow to run gradle check`

rishabh6788 replied:
Do you mean to say delete the uploaded artifact after benchmark job has been submitted?

reta replied:
Exactly, sorry about vague description

rishabh6788 replied:
I didn't add because it can take some time, upto 5-7 mins for the tarball to get downloaded for spinning up the cluster, if the jenkins agent nodes are not ready then even some more. 
However, I have added a lifecycle rule on the bucket to purge all the objects older than 30-days, I think I can reduce it even further to 1-day.

reta replied:
>  didn't add because it can take some time, upto 5-7 mins after the job has started, for the tarball to get downloaded for spinning up the cluster,

Ah I see, may be we could design the benchmarks the same way we do Gradle check by polling Jenkins URL? In this case it will be very clear when job is done (and exact moment when cleanup / purge should take place)

rishabh6788 replied:
I thought of that, but some workloads, like `big5` can run for 5-6 hours, `vectorsearch` indexing runs for 10-hrs, I am afraid we might hit the timeout for runners. 
I can take an action item to have this clean up done in the jenkins job as part of `post` action in the related jenkinsfile. 
Let me know your thoughts.

reta replied:
> I thought of that, but some workloads, like `big5` can run for 5-6 hours, `vectorsearch` indexing runs for 10-hrs, I am afraid we might hit the timeout for runners.

That is a good point

> I can take an action item to have this clean up done in the jenkins job as part of post action in the related jenkinsfile.
Let me know your thoughts.

I think N days purging ( lifecycle rule ) as you suggested before is the simplest way to go

rishabh6788 commented:
I have made the suggested changes in the jenkins job to limit concurrent runs to 20 and also added nightly runs for baseline comparison. I have also added required secrets in the OpenSearch repository for the workflow to work. 
Let me know if you have any other comments or concerns @reta

reta commented on .github/workflows/benchmark-pull-request.yml:
```diff
@@ -0,0 +1,142 @@
+              issue_number: context.issue.number,
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: 'Invalid comment format or config id. Please refer https://github.com/opensearch-project/OpenSearch/blob/main/PERFORMANCE_BENCHMARKS.md on how to run benchmarks on pull requests.'
```
Super nit:

```suggestion
              body: 'Invalid comment format or config id. Please refer to https://github.com/opensearch-project/OpenSearch/blob/main/PERFORMANCE_BENCHMARKS.md on how to run benchmarks on pull requests.'
```

reta commented on .github/workflows/benchmark-pull-request.yml:
```diff
@@ -0,0 +1,142 @@
+          secret: ${{ github.TOKEN }}
+          approvers: ${{ steps.get_approvers.outputs.approvers }}
+          minimum-approvals: 1
+          issue-title: 'Request to approve/deny benchmark for PR #${{ env.PR_NUMBER }}'
```
```suggestion
          issue-title: 'Request to approve/deny benchmark run for PR #${{ env.PR_NUMBER }}'
```

reta commented on .github/workflows/benchmark-pull-request.yml:
```diff
@@ -0,0 +1,142 @@
+          approvers: ${{ steps.get_approvers.outputs.approvers }}
+          minimum-approvals: 1
+          issue-title: 'Request to approve/deny benchmark for PR #${{ env.PR_NUMBER }}'
+          issue-body: "Please approve or deny the release for PR #${{ env.PR_NUMBER }}"
```
```suggestion
          issue-body: "Please approve or deny the benchmark run for PR #${{ env.PR_NUMBER }}"
```
(seems like copy/paste leftovers)

reta commented on .github/workflows/benchmark-pull-request.yml:
```diff
@@ -0,0 +1,142 @@
+      - name: Checkout opensearch-build repo
+        uses: actions/checkout@v4
+        with:
+          repository: rishabh6788/opensearch-build
```
Why personal repo here?

```suggestion
          repository: opensearch-project/opensearch-build
```

rishabh6788 replied:
oops!
Was supposed to revert this. Thank you for catching this.

reta commented:
> comparison. I have also added required secrets in the OpenSearch repository for the workflow to work.
> Let me know if you have any other comments or concerns @reta

LGTM, thanks @rishabh6788 , a few minor comments but good to otherwise, @peterzhuamazon mind please signing off as well since the build/ci infra is involved, thank you

peterzhuamazon reviewed (approved):
LGTM, Thanks!
